text/0030-protocol-messages.md

```diff
@@ -0,0 +1,246 @@
+* `[0b01101110, 0b01010001, 0b00010001]` denotes that the node supports protocol versions 2, 3, 4, 6, 7, 9, 13, 15, 17
+  and 21.
+
+### Version 1
+
+#### `LegacyGossip`
+
+Type ID: `2`
+
+A legacy message to broadcast a transaction and request another one at the same time.
+
+|Name|Description|Type|Length|
+|----|-----------|----|------|
+|`transaction`|Transaction to broadcast. Can be compressed<sup>1</sup>.|`Vec<u8>`|292-1604|
+|`hash`|Hash of the requested transaction.|`[u8; 49]`|49|
+
+<sup>1</sup> Compression is detailed at the end.
+
+**Note**: This message is the original IRI protocol message before the TLV protocol was introduced. It was kept by
+HORNET for compatibility with IRI but is not used between HORNET nodes. Its "ping-pong" concept has complex consequences
+on the node design and as such will not be implemented by Bee.
+
+### Version 2
+
+#### `MilestoneRequest`
+
+Type ID: `3`
+
+A message to request a milestone.
+
+|Name|Description|Type|Length|
+|----|-----------|----|------|
+|`index`|Index of the requested milestone.|`u32`|4|
+
+#### `TransactionBroadcast`
+
+Type ID: `4`
+
+A message to broadcast a transaction.
+
+|Name|Description|Type|Length|
+|----|-----------|----|------|
+|`transaction`|Transaction to broadcast. Can be compressed<sup>1</sup>.|`Vec<u8>`|292-1604|
+
+<sup>1</sup> Compression is detailed at the end.
+
+#### `TransactionRequest`
+
+Type ID: `5`
+
+A message to request a transaction.
+
+|Name|Description|Type|Length|
+|----|-----------|----|------|
+|`hash`|Hash of the requested transaction.|`[u8; 49]`|49|
+
+#### `Heartbeat`
+
+Type ID: `6`
+
+A message that informs about the part of the tangle currently being fully stored by a node.
+This message is sent when a node:
+* just got paired to another node;
+* did a local snapshot and pruned away a part of the tangle;
+* solidified a new milestone;
+
+It also helps other nodes to know if they can ask it a specific transaction.
+
+|Name|Description|Type|Length|
+|----|-----------|----|------|
+|`solid_milestone_index`|Index of the last solid milestone.|`u32`|4|
+|`snapshot_milestone_index`|Index of the snapshotted milestone.|`u32`|4|
+
+### Compression
+
+A transaction encoded in bytes has a length of `1604`. The `payload` field itself occupies `1312` bytes and is often
+partially or completely filled with `0`s. For this reason, trailing `0`s of the `payload` field are removed, providing
+a compression rate up to nearly 82%. Only the `payload` field is altered during this compression and the order of the
+fields stays the same.
+
+Proposed functions:
+```rust
+fn compress_transaction_bytes(bytes: &[u8]) -> Vec<u8> {
+    ...
+}
+
+fn uncompress_transaction_bytes(bytes: &[u8]) -> [u8; 1604] {
+    ...
+}
+```
+
+# Drawbacks
+
+Since IRI nodes only implement version `0` and `1` and Bee nodes only implement versions `0` and `2`, they will not be
+able to communicate with each other.
```

Comment:
Please mention, that this is only temporary until IRI implements STING.

Reply:
It's not temporary. IRI will never implement STING.